lib/sshutils/server.go

```diff
@@ -810,16 +837,43 @@ func (c *ConnectionWrapper) Read(b []byte) (int, error) {
 			skip = payloadBoundary + 1
 		}
 	}
-	c.upstreamReader = io.MultiReader(bytes.NewBuffer(buff[skip:]), c.Conn)
+	if bytes.HasPrefix(buff, multiplexer.ProxyV2Prefix) {
+		reader := bufio.NewReader(io.MultiReader(bytes.NewBuffer(buff), c.Conn))
+
+		newProxyLine, err := multiplexer.ReadProxyLineV2(reader)
+		if err != nil {
+			return 0, trace.Wrap(err)
+		}
+		if newProxyLine.IsSigned() && c.caGetter != nil {
+			err = newProxyLine.VerifySignature(context.Background(), c.caGetter, c.clusterName, c.clock)
+			if err != nil {
+				return 0, trace.Wrap(err)
+			}
+			if newProxyLine != nil {
+				c.clientAddr = &newProxyLine.Source
+			}
+		}
+
+		c.upstreamReader = reader
+	}
+
+	if c.upstreamReader == nil {
+		c.upstreamReader = io.MultiReader(bytes.NewBuffer(buff[skip:]), c.Conn)
+	}
 	return c.upstreamReader.Read(b)
 }
 
-// WrapConnection takes a network connection, wraps it into ConnectionWrapper
+type CertAuthorityGetter = func(ctx context.Context, id types.CertAuthID, loadKeys bool) (types.CertAuthority, error)
+
+// wrapConnection takes a network connection, wraps it into connectionWrapper
 // object (which overrides Read method) and returns the wrapper.
-func WrapConnection(conn net.Conn, logger logrus.FieldLogger) *ConnectionWrapper {
-	return &ConnectionWrapper{
-		Conn:       conn,
-		clientAddr: conn.RemoteAddr(),
-		logger:     logger,
+func wrapConnection(conn net.Conn, caGetter CertAuthorityGetter, localClusterName string, clock clockwork.Clock, logger logrus.FieldLogger) *connectionWrapper {
```

Comment:
question: should we enforce the `caGetter` to be non-nil ? Are there any scenarios where we won't need it?
If someone forgets to pass the `SetCaGetter` option to the SSHServer, we could skip the proxy header validation

Reply:
Server that is used in `reversetunnel` doesn't need it, since it doesn't get headers, connection already has corrected client address when server gets it.